reader.go

```diff
@@ -1,17 +1,361 @@
+	r.mutex.Lock()
+	if r.closed {
+		err = io.ErrClosedPipe
+	} else if offset != r.offset {
+		r.offset = offset
+		r.dirty = true
+		r.cancel() // may wake up a blocked Read call to spawn a new reader
+	}
+	r.mutex.Unlock()
+	return err
+}
+
+// A reader reads messages from kafka and produces them on its channels, it's
+// used as an way to asynchronously fetch messages while the main program reads
+// them using the high level reader API.
+type reader struct {
+	dialer    *Dialer
+	brokers   []string
+	topic     string
+	partition int
+	minBytes  int
+	maxBytes  int
+	maxWait   time.Duration
+	version   int64
+	msgs      chan<- readerMessage
+	errs      chan<- readerError
+}
+
+type readerMessage struct {
+	version int64
+	message Message
+}
+
+type readerError struct {
+	version int64
+	error   error
+}
+
+func (r *reader) run(ctx context.Context, offset int64) {
+	// This is the reader's main loop, it only ends if the context is canceled
+	// and will keep attempting to reader messages otherwise.
+	//
+	// Retrying indefinitely has the nice side effect of preventing Read calls
+	// on the parent reader to block if connection to the kafka server fails,
+	// the reader keeps reporting errors on the error channel which will then
+	// be surfaced to the program.
+	// If the reader wasn't retrying then the program would block indefinitely
+	// on a Read call after reading the first error.
+	for attempt := 0; true; attempt++ {
+		if attempt != 0 {
+			if !sleep(ctx, backoff(attempt, time.Second, time.Minute)) {
+				return
+			}
+		}
+
+		conn, start, err := r.initialize(ctx, offset)
+		if err != nil {
+			r.sendError(ctx, err)
+			continue
+		}
+
+		// Resetting the attempt counter ensures that if a failre occurs after
+		// a successful initialization we don't keep increasing the backoff
+		// timeout.
+		attempt = 0
+
+		// Now we're sure to have an absolute offset number, may anything happen
+		// to the connection we know we'll want to restart from this offset.
+		offset = start
+
+	readLoop:
+		for {
+			switch offset, err = r.read(ctx, offset, conn); err {
+			case nil:
+			case RequestTimedOut:
+			case context.Canceled: // another reader has taken over
+				conn.Close()
+				return
+			default:
+				conn.Close()
+				break readLoop
+			}
+		}
+	}
+}
+
+func (r *reader) initialize(ctx context.Context, offset int64) (conn *Conn, start int64, err error) {
+	var whence int
+
+	if offset < 0 {
+		offset = 0
+	} else {
+		whence = 1
+	}
+
+	for _, broker := range r.brokers {
+		if conn, err = r.dialer.DialLeader(ctx, "tcp", broker, r.topic, r.partition); err != nil {
+			continue
+		}
+		if start, err = conn.Seek(offset, whence); err != nil {
+			conn.Close()
+			break
+		}
+	}
+
+	return
+}
+
+func (r *reader) read(ctx context.Context, offset int64, conn *Conn) (int64, error) {
+	conn.SetReadDeadline(time.Now().Add(r.maxWait))
+	batch := conn.ReadBatch(r.minBytes, r.maxBytes)
+	conn.SetReadDeadline(time.Now().Add(10 * time.Second))
```

Comment:
It does, and it's the intent. When `ReadBatch` return the response has already been received and the header were read, so the initial `maxWait` value doesn't really matter anymore, I'm just trying to set a hard deadline on how long it may take to read the messages, just in case the connection gets aborted without the client knowing about it.